File: python/src/greybox/stick.py
"""Seasonality, Trend, and Irregular Contribution Kit (STI).

This module is a Python port of R's ``greybox::stick``. It decomposes the
variance of a time series into seasonal, trend and irregular parts based on
an Analysis of Variance (ANOVA) of the series on the seasonal and trend
factors and measures the contribution of each component.

The function implements the Seasonality, Trend, and Irregular (STI)
classification of Hans Levenbach (Levenbach, H. (2021). Four P's in a Pod:
e-Commerce Forecasting and Planning for Supply Chain Practitioners.
Independently published. ISBN 979-8461733575).
"""

from __future__ import annotations

import warnings

import numpy as np
import pandas as pd
from scipy import stats


class StickResult:
    """Result of :func:`stick`. Mirrors R's ``stick`` S3 class.

    Attributes
    ----------
    y : numpy.ndarray
        The original time series.
    lags : numpy.ndarray
        The seasonal lags used in the analysis (unique, ascending).
    anova : pandas.DataFrame
        The ANOVA table, with one row per seasonal lag, a ``trend`` row
        (when a trend was fitted) and a ``Residuals`` row. Columns are
        ``Df``, ``Sum Sq``, ``Mean Sq``, ``F value`` and ``Pr(>F)``.
    strength : pandas.Series
        The strength of each component: one entry per seasonal lag, the
        ``trend`` and the ``irregular`` component. The values are the
        shares of the respective Sum of Squares in the total Sum of
        Squares and sum up to one.
    """

    __slots__ = ("y", "lags", "anova", "strength")

    def __init__(
        self,
        y: np.ndarray,
        lags: np.ndarray,
        anova: pd.DataFrame,
        strength: pd.Series,
    ):
        self.y = y
        self.lags = lags
        self.anova = anova
        self.strength = strength

    def __str__(self) -> str:
        digits = 4
        lags_str = ", ".join(str(int(lag)) for lag in self.lags)
        rounded = self.strength.round(digits)
        return (
            "Seasonality, Trend, and Irregular Contribution Kit\n"
            f"Seasonal lags: {lags_str}\n\n"
            "Strength of the components:\n"
            f"{rounded.to_string()}"
        )

    def __repr__(self) -> str:
        lags_str = ", ".join(str(int(lag)) for lag in self.lags)
        return f"StickResult(lags=[{lags_str}])"

    def _reshape_to_matrix(self, period: int) -> np.ndarray:
        """Reshape ``y`` into a ``[period x cycle]`` matrix.

        Columns are cycles, rows are positions within the cycle. The last
        (incomplete) cycle is padded with NaNs.
        """
        y_values = np.asarray(self.y, dtype=float)
        nobs = len(y_values)
        n_cycles = int(np.ceil(nobs / period))
        padded = np.full(n_cycles * period, np.nan)
        padded[:nobs] = y_values
        # C-order reshape gives row i = cycle i; transpose so that the
        # columns are cycles and the rows are positions within the cycle.
        return padded.reshape(n_cycles, period).T

    def _plot_seasonal(self, ax, period: int, **kwargs) -> None:
        """Draw the seasonal plot for one lag on the given axes."""
        matrix = self._reshape_to_matrix(period)
        n_row, n_col = matrix.shape
        x_idx = np.arange(1, n_row + 1)
        # One grey line per cycle, light-to-dark gradient.
        shades = 0.85 * np.arange(1, n_col + 1) / n_col
        for j in range(n_col):
            ax.plot(x_idx, matrix[:, j], color=str(shades[j]), **kwargs)
        # The average seasonal profile across the cycles.
        ax.plot(
            x_idx,
            np.nanmean(matrix, axis=1),
            color="black",
            linewidth=2,
            linestyle="--",
        )
        ax.set_title(f"Seasonal plot (lag {int(period)})")
        ax.set_xlabel("Period within the cycle")
        ax.set_ylabel("Value")

    def _plot_trend(self, ax, period: int, **kwargs) -> None:
        """Draw the trend plot (uses the longest lag) on the given axes."""
        matrix = self._reshape_to_matrix(period)
        n_row, n_col = matrix.shape
        x_idx = np.arange(1, n_col + 1)
        # One grey line per seasonal position, drawn across the cycles.
        shades = 0.85 * np.arange(1, n_row + 1) / n_row
        for i in range(n_row):
            ax.plot(x_idx, matrix[i, :], color=str(shades[i]), **kwargs)
        # The average level per cycle: the trend.
        ax.plot(
            x_idx,
            np.nanmean(matrix, axis=0),
            color="black",
            linewidth=2,
            linestyle="--",
        )
        ax.set_title("Trend plot")
        ax.set_xlabel("Cycle")
        ax.set_ylabel("Value")

    def plot(self, which=None, axes=None, **kwargs):
        """Plot the seasonal and trend components.

        Mirrors R's ``plot.stick()``. For every seasonal lag a "seasonal
        plot" is produced (the series reshaped into one grey line per
        cycle, with the average seasonal profile overlaid as a bold dashed
        black line); the final plot is the "trend plot" (one grey line per
        seasonal position drawn across the cycles, with the average level
        per cycle -- the trend -- overlaid as a bold dashed black line).

        Parameters
        ----------
        which : int or sequence of int, optional
            Which plots to produce: ``1, ..., k-1`` correspond to the
            seasonal plots for each of the seasonal lags (in ascending
            order), while ``k`` corresponds to the trend plot (the last
            one). If ``None`` (the default), all the plots are produced.
            Values outside the valid range are dropped with a warning.
        axes : matplotlib.axes.Axes or sequence of Axes, optional
            Axes to draw on, one per requested plot. A new figure with a
            row of subplots is created if ``None``.
        **kwargs
            Forwarded to :meth:`matplotlib.axes.Axes.plot` for the grey
            component lines.

        Returns
        -------
        numpy.ndarray of matplotlib.axes.Axes
            The axes containing the plots.
        """
        import matplotlib.pyplot as plt

        n_lags = len(self.lags)
        has_trend = "trend" in self.strength.index
        n_plots = n_lags + (1 if has_trend else 0)

        if which is None:
            which = list(range(1, n_plots + 1))
        elif np.isscalar(which):
            which = [int(which)]
        else:
            which = [int(w) for w in which]

        valid = [w for w in which if 1 <= w <= n_plots]
        if len(valid) != len(which):
            warnings.warn(
                f"Some values of 'which' are outside the range "
                f"[1, {n_plots}] and are dropped.",
                stacklevel=2,
            )
        which = valid
        if len(which) == 0:
            return np.array([], dtype=object)

        if axes is None:
            _, axes = plt.subplots(1, len(which))
        axes = np.atleast_1d(axes)

        lag_max = int(np.max(self.lags))
        for ax, w in zip(axes, which):
            if w <= n_lags:
                self._plot_seasonal(ax, int(self.lags[w - 1]), **kwargs)
            else:
                self._plot_trend(ax, lag_max, **kwargs)
        return axes


def _residual_sum_of_squares(y: np.ndarray, design: np.ndarray) -> float:
    """Residual sum of squares of an OLS fit of ``y`` on ``design``."""
    coef, _, _, _ = np.linalg.lstsq(design, y, rcond=None)
    residuals = y - design @ coef
    return float(residuals @ residuals)


def _factor_dummies(codes: np.ndarray) -> np.ndarray:
    """Treatment-coded dummies for a factor (first level dropped)."""
    levels = np.unique(codes)
    # Drop the first level for identifiability (treatment contrasts).
    return np.column_stack([(codes == level).astype(float) for level in levels[1:]])


def stick(y, lags=None) -> StickResult:
    """Seasonality, Trend, and Irregular Contribution Kit.

    Decomposes the variance of a time series into seasonal, trend and
    irregular parts based on an Analysis of Variance (ANOVA) of the series
    on the seasonal and trend factors, and measures the contribution of
    each component. This is a Python port of R's :func:`greybox::stick`.

    A data frame is formed internally with the response ``y``, a
    categorical (factor) variable for each of the provided seasonal
    ``lags`` and a "trend" factor. For a monthly series with ``lags=12``,
    the seasonal factor takes values ``1..12`` repeated throughout the
    sample (the month of the year), while the trend factor takes values
    ``1..ceil(T/12)``, each repeated 12 times (the year). The trend factor
    is constructed based on the longest of the provided lags. An ANOVA is
    then applied to ``y ~ seasonal + trend`` and the strength of each
    component is measured as the share of the respective Sum of Squares in
    the total Sum of Squares. The irregular component corresponds to the
    share of the residual Sum of Squares; the strengths sum up to one.

    The function implements the Seasonality, Trend, and Irregular (STI)
    classification of Hans Levenbach.

    Parameters
    ----------
    y : array-like
        The time series to analyse.
    lags : int or sequence of int
        The seasonal lags (periodicities) in the data, e.g.
        ``lags=[24, 168]`` for hourly data or ``lags=12`` for monthly
        data. Values not greater than one are dropped.

    Returns
    -------
    StickResult
        Object with the original data, the ``lags`` used, the ANOVA table
        and the ``strength`` of each component.

    References
    ----------
    Levenbach, H. (2021). Four P's in a Pod: e-Commerce Forecasting and
    Planning for Supply Chain Practitioners. Independently published.
    ISBN 979-8461733575.

    Examples
    --------
    >>> import numpy as np
    >>> t = np.arange(1, 121)
    >>> y = 100 + 0.3 * t + 20 * np.sin(2 * np.pi * t / 12)
    >>> result = stick(y, lags=12)
    >>> bool(np.isclose(result.strength.sum(), 1.0))
    True
    """
    if lags is None:
        raise ValueError("'lags' must be provided (e.g. lags=12).")

    y_values = np.asarray(y, dtype=float).ravel()
    nobs = len(y_values)

    # Only seasonal lags above one make sense; keep unique and sorted.
    lags = np.atleast_1d(np.asarray(lags))
    lags = np.unique(lags[lags > 1]).astype(int)
    if len(lags) == 0:
        raise ValueError(
            "No seasonal lags greater than one were provided. "
            "stick() needs at least one seasonal periodicity."
        )

    # Drop the lags that are too long to form at least two full cycles.
    too_long = lags >= nobs
    if np.any(too_long):
        warnings.warn(
            "The following lags are not smaller than the number of "
            "observations and are dropped: "
            f"{', '.join(str(int(lag)) for lag in lags[too_long])}.",
            stacklevel=2,
        )
        lags = lags[~too_long]
    if len(lags) == 0:
        raise ValueError("None of the provided lags is shorter than the sample.")

    positions = np.arange(nobs)

    # Build the ordered list of terms: a seasonal factor per lag, then the
    # trend ("year") factor based on the longest lag.
    term_names = [f"seasonal{int(lag)}" for lag in lags]
    term_codes = [positions % int(lag) for lag in lags]

    lag_max = int(np.max(lags))
    trend_codes = positions // lag_max
    has_trend = len(np.unique(trend_codes)) > 1
    if has_trend:
        term_names.append("trend")
        term_codes.append(trend_codes)

    # Sequential (Type I) ANOVA: add each term's dummies in order and
    # record the reduction in the residual sum of squares.
    intercept = np.ones((nobs, 1))
    design = intercept
    rss_prev = _residual_sum_of_squares(y_values, design)

    df = []
    sum_sq = []
    for codes in term_codes:
        dummies = _factor_dummies(codes)
        design = np.column_stack([design, dummies])
        rss_curr = _residual_sum_of_squares(y_values, design)
        df.append(dummies.shape[1])
        sum_sq.append(rss_prev - rss_curr)
        rss_prev = rss_curr

    # Residual ("irregular") row.
    df_residual = nobs - design.shape[1]
    sum_sq_residual = rss_prev

    rows = term_names + ["Residuals"]
    df_all = np.array(df + [df_residual], dtype=float)
    sum_sq_all = np.array(sum_sq + [sum_sq_residual], dtype=float)
    mean_sq = sum_sq_all / df_all

    # F values and p-values for the model terms (not for the residuals).
    mean_sq_residual = mean_sq[-1]
    f_value = np.full(len(rows), np.nan)
    p_value = np.full(len(rows), np.nan)
    if df_residual > 0:
        f_value[:-1] = mean_sq[:-1] / mean_sq_residual
        p_value[:-1] = stats.f.sf(f_value[:-1], df_all[:-1], df_residual)

    anova = pd.DataFrame(
        {
            "Df": df_all,
            "Sum Sq": sum_sq_all,
            "Mean Sq": mean_sq,
            "F value": f_value,
            "Pr(>F)": p_value,
        },
        index=rows,
    )

    # Strength of each component = its share of the total Sum of Squares.
    strength_values = sum_sq_all / sum_sq_all.sum()
    strength_names = term_names + ["irregular"]
    strength = pd.Series(strength_values, index=strength_names)

    return StickResult(y=y_values, lags=lags, anova=anova, strength=strength)

File: R/stick.R
#' Seasonality, Trend, and Irregular Contribution Kit
#'
#' The function decomposes the variance of a time series into seasonal, trend
#' and irregular parts based on an Analysis of Variance (ANOVA) of the series
#' on the seasonal and trend factors and measures their contribution to the data.
#'
#' \code{stick()} forms internally a data frame with the response \code{y} in
#' the first column, a categorical (factor) variable for each of the provided
#' seasonal \code{lags} and a "trend" factor in the last column. For a monthly
#' series with \code{lags=12}, the seasonal factor takes values \code{1:12}
#' repeated throughout the sample (the month of the year), while the trend
#' factor takes values \code{1:ceiling(T/12)}, each repeated 12 times (the
#' year). The trend factor is constructed based on the longest of the provided
#' lags.
#'
#' \code{aov()} is then applied to the model \code{y ~ seasonal + trend} and the
#' strength of each component is measured as the share of the respective Sum of
#' Squares in the total Sum of Squares. The irregular component corresponds to
#' the share of the residual Sum of Squares. The strengths of all the components
#' sum up to one.
#'
#' The function implements the Seasonality, Trend, and Irregular (STI)
#' classification of Hans Levenbach (see the reference below).
#'
#' @param y The vector or a ts object, containing the time series to analyse.
#' @param lags The vector of seasonal lags (periodicities) in the data, e.g.
#' \code{lags=c(24,168)} for the hourly data or \code{lags=12} for the monthly
#' one. Values not greater than one are dropped. Defaults to the frequency of
#' \code{y}.
#' @param ... Other parameters passed to the \code{aov()} function.
#'
#' @return \code{stick()} returns an object of class "stick", which contains:
#' \itemize{
#' \item y - The original data;
#' \item lags - The seasonal lags used in the analysis;
#' \item model - The fitted \code{aov} model;
#' \item anova - The ANOVA table (the first element of \code{summary(model)});
#' \item strength - The named vector with the strength of each seasonal
#' component, the trend and the irregular component. The values sum up to one.
#' }
#'
#' @template author
#' @template keywords
#'
#' @references Levenbach, H. (2021). Four P's in a Pod: e-Commerce Forecasting
#' and Planning for Supply Chain Practitioners. Independently published.
#' ISBN 979-8461733575.
#'
#' @seealso \code{\link[stats]{aov}}
#'
#' @examples
#' # Generate a seasonal series and classify it
#' y <- ts(rnorm(120, 100 + 20 * sin(2 * pi * (1:120) / 12), 5), frequency=12)
#' stick(y)
#'
#' @importFrom stats aov as.formula frequency
#' @export
stick <- function(y, lags=frequency(y), ...){
    obsInSample <- length(y);

    # Only seasonal lags above one make sense; keep unique and sorted
    lags <- sort(unique(lags[lags>1]));
    if(length(lags)==0){
        stop(paste0("No seasonal lags greater than one were provided. ",
                    "stick() needs at least one seasonal periodicity."),
             call.=FALSE);
    }

    # Drop the lags that are too long to form at least two full cycles
    lagsTooLong <- lags>=obsInSample;
    if(any(lagsTooLong)){
        warning(paste0("The following lags are not smaller than the number of ",
                       "observations and are dropped: ",
                       paste(lags[lagsTooLong], collapse=", "), "."),
                call.=FALSE);
        lags <- lags[!lagsTooLong];
    }
    if(length(lags)==0){
        stop("None of the provided lags is shorter than the sample.", call.=FALSE);
    }

    # The data frame with y in the first column
    sticksData <- data.frame(y=as.vector(y));

    # A seasonal factor for each lag: the position within the cycle
    seasonalNames <- paste0("seasonal", lags);
    for(i in seq_along(lags)){
        sticksData[[seasonalNames[i]]] <-
            factor(((seq_len(obsInSample)-1) %% lags[i]) + 1);
    }

    # The trend ("year") factor, based on the longest lag
    lagMax <- max(lags);
    trendFactor <- factor(((seq_len(obsInSample)-1) %/% lagMax) + 1);
    # Only include the trend if it has at least two levels
    trendIncluded <- nlevels(trendFactor)>1;
    terms <- seasonalNames;
    if(trendIncluded){
        sticksData[["trend"]] <- trendFactor;
        terms <- c(terms, "trend");
    }

    # Build the formula and fit the ANOVA model
    sticksFormula <- as.formula(paste0("y ~ ", paste(terms, collapse=" + ")));
    sticksModel <- aov(sticksFormula, data=sticksData, ...);
    sticksANOVA <- summary(sticksModel)[[1]];

    # The strength of each component is its share of the total Sum of Squares
    SSQ <- sticksANOVA[,"Sum Sq"];
    strength <- SSQ / sum(SSQ);
    names(strength) <- trimws(rownames(sticksANOVA));
    # The residual Sum of Squares corresponds to the irregular component
    names(strength)[names(strength)=="Residuals"] <- "irregular";

    return(structure(list(y=y, lags=lags, model=sticksModel, anova=sticksANOVA,
                          strength=strength),
                     class="stick"));
}

#' @param x The object of class "stick" generated by the \code{stick()} function.
#' @param digits The number of digits to use when printing the strength of the
#' components.
#' @rdname stick
#' @export
print.stick <- function(x, digits=4, ...){
    cat("Seasonality, Trend, and Irregular Contribution Kit\n");
    cat("Seasonal lags: ", paste(x$lags, collapse=", "), "\n\n", sep="");
    cat("Strength of the components:\n");
    print(round(x$strength, digits));
}

#' @param which The vector of integers specifying which plots to produce: the
#' values \code{1, ..., k-1} correspond to the seasonal plots for each of the
#' seasonal \code{lags} (in the ascending order), while \code{k} corresponds to
#' the trend plot (the last one). If \code{NULL} (the default), all the plots
#' are produced.
#' @param ask Logical; if \code{TRUE}, the user is asked to hit Enter before
#' each new plot is drawn. The default only asks when the requested plots do not
#' all fit on the current device layout (see \link[graphics]{par}).
#'
#' @details
#' \code{plot.stick()} reproduces two types of plots, in the spirit of the STI
#' decomposition. For every seasonal lag it draws a "seasonal plot": the series
#' is reshaped into a matrix with one column per cycle (e.g. per year for
#' monthly data) and the value within the cycle on the x-axis; each cycle is a
#' grey line and the average seasonal profile (the mean across the cycles) is
#' overlaid as a bold dashed black line. The final plot is the "trend plot",
#' which uses the longest lag: each seasonal position (e.g. each month) is a
#' grey line drawn across the cycles, and the average level per cycle (the
#' trend) is overlaid as a bold dashed black line.
#'
#' @return \code{plot.stick()} is called for its side effect of producing the
#' plots and returns \code{NULL} invisibly.
#'
#' @examples
#' # Produce all the plots
#' par(mfcol=c(1,2))
#' plot(stick(AirPassengers))
#' # Only the trend plot
#' plot(stick(AirPassengers), which=2)
#'
#' @importFrom grDevices grey dev.interactive devAskNewPage
#' @importFrom graphics plot lines par
#' @rdname stick
#' @export
plot.stick <- function(x, which=NULL, ask, ...){
    lags <- x$lags;
    nLags <- length(lags);
    # The trend plot is only available if stick() actually fitted a trend
    hasTrend <- "trend" %in% names(x$strength);
    nPlots <- nLags + as.integer(hasTrend);

    # Resolve which plots to produce
    if(is.null(which)){
        which <- seq_len(nPlots);
    }
    else{
        whichValid <- which %in% seq_len(nPlots);
        if(any(!whichValid)){
            warning(paste0("Some values of 'which' are outside the range [1, ",
                           nPlots, "] and are dropped."), call.=FALSE);
            which <- which[whichValid];
        }
    }
    if(length(which)==0){
        return(invisible(NULL));
    }

    # Wait for "Enter" between plots only if they do not fit the layout,
    # mirroring the behaviour of plot.greybox().
    if(missing(ask)){
        ask <- prod(par("mfcol")) < length(which) && dev.interactive();
    }
    if(ask){
        oask <- devAskNewPage(TRUE);
        on.exit(devAskNewPage(oask));
    }

    yValues <- as.vector(x$y);
    obsInSample <- length(yValues);
    lagMax <- max(lags);

    # Reshape the series into a [period x cycle] matrix, padding the last
    # (incomplete) cycle with NAs. Columns are cycles, rows are the positions
    # within the cycle.
    reshapeToMatrix <- function(period){
        nCycles <- ceiling(obsInSample / period);
        yPadded <- c(yValues, rep(NA, nCycles*period - obsInSample));
        matrix(yPadded, nrow=period, ncol=nCycles);
    }

    # Seasonal plot: one grey line per cycle, the average seasonal profile
    # (mean across the cycles) overlaid as a bold dashed black line.
    plotSeasonal <- function(seasonalMatrix, lag, ...){
        ellipsis <- list(...);
        nRow <- nrow(seasonalMatrix);
        nCol <- ncol(seasonalMatrix);

        if(is.null(ellipsis$main)){
            ellipsis$main <- paste0("Seasonal plot (lag ", lag, ")");
        }
        if(is.null(ellipsis$xlab)){
            ellipsis$xlab <- "Period within the cycle";
        }
        if(is.null(ellipsis$ylab)){
            ellipsis$ylab <- "Value";
        }
        if(is.null(ellipsis$ylim)){
            ellipsis$ylim <- range(seasonalMatrix, na.rm=TRUE);
        }
        cols <- grey(0.85 * seq_len(nCol) / nCol);

        ellipsis$x <- seq_len(nRow);
        ellipsis$y <- seasonalMatrix[,1];
        ellipsis$type <- "l";
        ellipsis$col <- cols[1];
        do.call(plot, ellipsis);
        for(j in seq_len(nCol)[-1]){
            lines(seq_len(nRow), seasonalMatrix[,j], col=cols[j]);
        }
        # The average seasonal profile across the cycles
        lines(seq_len(nRow), rowMeans(seasonalMatrix, na.rm=TRUE),
              col="black", lwd=2, lty=2);
    }

    # Trend plot: one grey line per seasonal position across the cycles, the
    # average level per cycle (the trend) overlaid as a bold dashed black line.
    plotTrend <- function(seasonalMatrix, ...){
        ellipsis <- list(...);
        nRow <- nrow(seasonalMatrix);
        nCol <- ncol(seasonalMatrix);

        if(is.null(ellipsis$main)){
            ellipsis$main <- "Trend plot";
        }
        if(is.null(ellipsis$xlab)){
            ellipsis$xlab <- "Cycle";
        }
        if(is.null(ellipsis$ylab)){
            ellipsis$ylab <- "Value";
        }
        if(is.null(ellipsis$ylim)){
            ellipsis$ylim <- range(seasonalMatrix, na.rm=TRUE);
        }
        cols <- grey(0.85 * seq_len(nRow) / nRow);

        ellipsis$x <- seq_len(nCol);
        ellipsis$y <- seasonalMatrix[1,];
        ellipsis$type <- "l";
        ellipsis$col <- cols[1];
        do.call(plot, ellipsis);
        for(i in seq_len(nRow)[-1]){
            lines(seq_len(nCol), seasonalMatrix[i,], col=cols[i]);
        }
        # The average level per cycle: the trend
        lines(seq_len(nCol), colMeans(seasonalMatrix, na.rm=TRUE),
              col="black", lwd=2, lty=2);
    }

    for(i in which){
        if(i<=nLags){
            plotSeasonal(reshapeToMatrix(lags[i]), lags[i], ...);
        }
        else{
            plotTrend(reshapeToMatrix(lagMax), ...);
        }
    }

    return(invisible(NULL));
}
